### PR TITLE
Post and Profile Images

### PR DESCRIPTION
**Pull request by john-royal**
This modifies `Tim-Images` to resolve merge conflicts with `main` and make some other improvements.

Image uploads have been reworked to:
* integrate better with rest of `UserService` and `PostService`
* improve UI/UX
* ensure errors are handled
* streamline codebase

**Update:** Changed base branch to `main` because the original PR for user & post images, #22, has been closed.

### CODE REVIEW

**Comment by Evilsmaher** [comment]
@john-royal Quick question -- for `unfavorite` inside **PostService.swift** -- can we not assume that the `snapshot` will return at most 1 item? Curious to understand the iteration there.

**Comment by john-royal** [comment]
Hi @Evilsmaher, this is more related to #18 than to this thread, but to answer your question, I used iteration because the `getDocuments` call returns an array of documents, as opposed to a single document.

While our assumption _should_ be correct, there are circumstances under which it won’t. For example:
* bugs in the learner’s code
* client is out of sync with backend and calls `favorite(_:)` more than once
* data is modified manually through the Firebase console

That’s why I’d prefer to trust the Swift type system over our own assumptions. It’s an edge case, but Swift is designed to encourage explicit handling of those edge cases (e.g. by removing an item from an array using `removeAll(where:)` instead of `remove(at:)`) because it helps us write more robust, less buggy code.

That's my opinion—please let me know if you disagree. Might this be an instance of overly defensive programming? Would a simpler solution (e.g. assertions) be better?

File: FirebaseTestProject/Views/NewPostForm.swift [hunk 1]
```diff
@@ -10,31 +10,73 @@ import SwiftUI
...
+
     
     var body: some View {
         Form {
             TextField("Title", text: $title)
                 .focused($showingKeyboard)
-            TextField("Post content", text: $postContent)
+            TextEditor(text: $postContent)
                 .focused($showingKeyboard)
+                .multilineTextAlignment(/*@START_MENU_TOKEN@*/.leading/*@END_MENU_TOKEN@*/)
```
Comment (benstone1) [inline FirebaseTestProject/Views/NewPostForm.swift:29]: We should remove this tokens

File: FirebaseTestProject/Views/NewPostForm.swift [hunk 2]
```diff
@@ -10,31 +10,73 @@ import SwiftUI
...
+                .frame(width: 300, height: 300, alignment: .topLeading)
+                
+            if let image = image {
+                Image(uiImage: image)
+                    .resizable()
+                    .aspectRatio(contentMode: .fit)
+                    .frame(width: 200, height: 200, alignment: .center)
+            } else {
+                Image(systemName: "photo")
+                    .resizable()
```
Comment (benstone1) [inline FirebaseTestProject/Views/NewPostForm.swift:39]: Can these attributes be consolidated?  Would be nice to not need to write out the same ones twice.

File: FirebaseTestProject/Views/PostRow.swift [hunk 3]
```diff
@@ -40,6 +43,24 @@ struct PostRow: View {
         }
     }
     
+    private var image: some View {
```
Comment (benstone1) [inline FirebaseTestProject/Views/PostRow.swift:46]: It might make more sense for this to be a separate structure instead of a property.  Then the semantics are a little nice for adding it:

```
VStack {
    Text("Hi")
    image
    Text("Bye")
}
```

is a little less SwiftUI-looking than

```
VStack {
    Text("Hi")
    PostImage(url)
    Text("Bye")
}
```

File: FirebaseTestProject/Views/PostRow.swift [hunk 4]
```diff
@@ -40,6 +43,24 @@ struct PostRow: View {
...
+                ProgressView()
+                Text("Loading Image")
+                    .font(.caption)
+            }
+            .frame(width: 300, height: 200)
+        }
+        .padding(.horizontal)
+    }
+    
     private var footer: some View {
```
Comment (benstone1) [inline FirebaseTestProject/Views/PostRow.swift:64]: Same for the footer

File: FirebaseTestProject/Views/NewPostForm.swift [hunk 5]
```diff
@@ -10,31 +10,73 @@ import SwiftUI
...
+                    }  label: {
+                        Label("Take Photo", systemImage: "camera")
+                    }
+                }
+                .buttonStyle(.borderless)
+                .labelStyle(.iconOnly)
+            }
             Button("Submit", action: submitPost)
         }
         .alert("Cannot Submit Post", isPresented: $submitTask.isError, presenting: submitTask.error) { error in
```
Comment (benstone1) [inline FirebaseTestProject/Views/NewPostForm.swift:63]: Does `isError` ever reset in the ViewModel?  Not part of this PR, just curious.

File: FirebaseTestProject/Views/ProfileView.swift [hunk 6]
```diff
@@ -40,9 +66,29 @@ struct ProfileView: View {
             .disabled(signOutTask.isInProgress)
             Spacer()
         }
+        .alert("Cannot Update Profile Photo", isPresented: $updateImageTask.isError, presenting: updateImageTask.error, actions: { _ in }) { error in
+            Text(error.localizedDescription)
+        }
         .alert("Cannot Sign Out", isPresented: $signOutTask.isError, presenting: signOutTask.error, actions: { _ in }) { error in
```
Comment (benstone1) [inline FirebaseTestProject/Views/ProfileView.swift:72]: Multiple alert modifiers is an interesting pattern.  Is there a way to consolidate these, or is this an established practice?

**Review by benstone1 — CHANGES_REQUESTED** [review]
Looks good, just added a few suggestions/questions/comments inside.

File: FirebaseTestProject/ViewModels/AuthViewModel.swift [hunk 7]
```diff
@@ -33,4 +34,8 @@ import Foundation
         try userService.signOut()
         user = nil
     }
+    
+    func updateImage(_ image: UIImage) async throws {
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/ViewModels/AuthViewModel.swift:34]: I think so! I'll update it. Should I do the same for the `updateImage(_:for:)` in the `UserService`?

File: FirebaseTestProject/ViewModels/AuthViewModel.swift [hunk 8]
```diff
@@ -33,4 +34,8 @@ import Foundation
...
+        user = try await userService.updateImage(image, for: user!)
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/ViewModels/AuthViewModel.swift:35]: Got it.

File: FirebaseTestProject/Services/PostService.swift [hunk 9]
```diff
@@ -76,9 +92,9 @@ struct PostService {
 private struct Favorite: Identifiable, FirebaseConvertable {
     let id: UUID
     let postid: UUID
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/PostService.swift]: Fixed!

File: FirebaseTestProject/Views/NewPostForm.swift (hunk 5 above)
Comment (john-royal) [reply FirebaseTestProject/Views/NewPostForm.swift:54]: It resets, but not within the view model. Since it’s passed as an `isPresented` binding, it gets updated automatically when the alert is dismissed.

File: FirebaseTestProject/Views/PostRow.swift (hunk 3 above)
Comment (john-royal) [reply FirebaseTestProject/Views/PostRow.swift]: I agree! I updated the view just now.

File: FirebaseTestProject/Views/NewPostForm.swift (hunk 2 above)
Comment (john-royal) [reply FirebaseTestProject/Views/NewPostForm.swift]: Done!

File: FirebaseTestProject/Views/NewPostForm.swift (hunk 1 above)
Comment (john-royal) [reply FirebaseTestProject/Views/NewPostForm.swift]: Nice catch! Fixed just now

File: FirebaseTestProject/Views/ImagePickerView.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+import UIKit
+import SwiftUI
+
+struct ImagePickerView: UIViewControllerRepresentable {
+    var sourceType: SourceType
+    @Binding var selection: UIImage?
+    @Environment(\.dismiss) private var dismiss
+    
+    typealias SourceType = UIImagePickerController.SourceType
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Views/ImagePickerView.swift]: Good idea! Updated just now

File: FirebaseTestProject/Views/PostRow.swift (hunk 4 above)
Comment (john-royal) [reply FirebaseTestProject/Views/PostRow.swift]: For this one, I wasn’t sure about moving the entire footer to a separate struct (a footer struct would require 3-4 parameters and be tightly coupled to the parent `PostRow`, which isn’t ideal). Instead, I separated out the `FavoriteButton` and `DeleteButton`—including their associated `TaskViewModel`s, alerts, and confirmation dialogs—and kept the remainder in the view `body`. Not sure if this was the best way to do it. If you could take a look and let me know what you think, I would really appreciate it!

File: FirebaseTestProject/Views/ProfileView.swift (hunk 6 above)
Comment (john-royal) [reply FirebaseTestProject/Views/ProfileView.swift:72]: This isn’t an established practice AFAIK, so consolidating these would be a good idea.

The simplest way to consolidate these would be to create one catch-all alert. This would mean the alert title would have to the same for both profile photo and authentication errors (unless we’re willing to use the [now-deprecated alert API](https://developer.apple.com/documentation/swiftui/view/alert(ispresented:content:))).

The alternative, which I think is preferable, would be the [new `alert(isPresented:error:actions:message:)` modifier](https://developer.apple.com/documentation/swiftui/form/alert(ispresented:error:actions:message:)). To use that modifier, we’d need to create custom error types that conform to `LocalizedError` (the `errorDescription` property becomes the alert title).

Please let me know what you think!

File: FirebaseTestProject/Models/User.swift [hunk 11]
```diff
@@ -8,13 +8,15 @@
 import Foundation
 
 struct User: Equatable, FirebaseConvertable {
-    let id: UUID
+    let id: String
+    var imageURL: String
     var name: String
     
-    init(id: UUID = .init(), name: String) {
+    init(id: String, name: String, imageURL: String = "https://firebasestorage.googleapis.com/v0/b/instacademy-test1.appspot.com/o/images%2Fusers%2Fdefault_profile.png?alt=media&token=6700ffd7-2675-45ad-a197-bb2e4e307da7") {
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Models/User.swift]: Good idea! I would implement this right away, but I'm unsure of how to reference an image from the app bundle from within the `User` model, since the model stores a URL. I know URLs can reference local files, but if a local file URL from one device is stored in Firebase, would it work on other devices as well?

[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Models/User.swift]: I believe the `Bundle` has a [`url(forResource:withExtension:)`](https://developer.apple.com/documentation/foundation/bundle/1411540-url) method that gives you the specific URL on a user's device for `defaultPhoto.png` or whatever.   I think that would probably work?

File: FirebaseTestProject/ViewModels/AuthViewModel.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/ViewModels/AuthViewModel.swift]: Probably, yeah, makes it more clear what image is changing.

File: FirebaseTestProject/Views/PostRow.swift (hunk 3 above)
Comment (benstone1) [reply FirebaseTestProject/Views/PostRow.swift]: Great, there's a couple more of these lurking in the codebase, we'll knock them out as we see them.

File: FirebaseTestProject/Views/ProfileView.swift (hunk 6 above)
Comment (benstone1) [reply FirebaseTestProject/Views/ProfileView.swift:72]: This does seem like a better approach, yeah.  Let's see what this would look like.  A single alert title is fine I think.

File: FirebaseTestProject/Models/User.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Models/User.swift]: It should work! I was skeptical about making it part of the `User` model since we don’t want a local file URL to be saved to Firebase by accident, but I doubt that’ll happen. I’ll push that change in a few minutes!

File: FirebaseTestProject/ViewModels/AuthViewModel.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/ViewModels/AuthViewModel.swift]: Done!

File: FirebaseTestProject/Views/ProfileView.swift (hunk 6 above)
Comment (john-royal) [reply FirebaseTestProject/Views/ProfileView.swift]: I tried consolidating the alerts. Does this work?

File: FirebaseTestProject/Views/ImagePickerView.swift [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    func makeUIViewController(context: Context) -> UIImagePickerController {
+        let imagePicker = UIImagePickerController()
+        imagePicker.sourceType = sourceType
+        imagePicker.delegate = context.coordinator
+        imagePicker.allowsEditing = true
+        return imagePicker
+    }
+    
+    func updateUIViewController(_ uiViewController: UIImagePickerController, context: Context) {
+
```
Comment (benstone1) [inline FirebaseTestProject/Views/ImagePickerView.swift:31]: Do we need this for `UIViewControllerRepresentable` conformance?

File: FirebaseTestProject/Views/NewPostForm.swift [hunk 13]
```diff
@@ -10,31 +10,71 @@ import SwiftUI
...
+        .sheet(item: $imageSourceType) {
+            ImagePickerView(sourceType: $0, selection: $image)
+        }
+    }
+    
+    private var uploadedImageOrPlaceholder: Image {
+        if let image = image {
+            return Image(uiImage: image)
+        }
+        return Image(systemName: "photo")
```
Comment (benstone1) [inline FirebaseTestProject/Views/NewPostForm.swift:66]: Super minor style thing, but the last return might be better in an `else` block.

File: FirebaseTestProject/Views/PostRow.swift [hunk 14]
```diff
@@ -14,59 +14,97 @@ struct PostRow: View {
...
+                task.request(with: action)
+            } label: {
+                Label("Delete", systemImage: "trash")
+                    .foregroundColor(.red)
+            }
+            .disabled(task.isInProgress)
+            .confirmationDialog("Are you sure you want to delete this post?", isPresented: $task.isPending, titleVisibility: .visible, presenting: task.confirmAction) {
+                Button("Delete", role: .destructive, action: $0)
+            }
+            .alert("Cannot Delete Post", isPresented: $task.isError, presenting: task.error) { error in
```
Comment (benstone1) [inline FirebaseTestProject/Views/PostRow.swift:104]: Consolidated alerts look good 👍

File: FirebaseTestProject/Views/ProfileView.swift [hunk 15]
```diff
@@ -37,21 +56,38 @@ struct ProfileView: View {
...
             }
-            .disabled(signOutTask.isInProgress)
+            .disabled(task.isInProgress)
             Spacer()
         }
-        .alert("Cannot Sign Out", isPresented: $signOutTask.isError, presenting: signOutTask.error, actions: { _ in }) { error in
+        .alert("Error", isPresented: $task.isError, presenting: task.error, actions: { _ in }) { error in
             Text(error.localizedDescription)
         }
+        .alert("Choose Profile Photo", isPresented: $showChooseImageSource) { // When the .confirmationDialog() modifier was used instead of an alert, the other alerts became inoperable (probably a SwiftUI bug), meaning that any potential errors were not propagated to the user.
```
Comment (benstone1) [inline FirebaseTestProject/Views/ProfileView.swift:65]: Could these not be combined?  Might have missed a note on it.

File: FirebaseTestProject/Models/Post.swift [hunk 16]
```diff
@@ -13,21 +13,31 @@ struct Post: Identifiable, Equatable, FirebaseConvertable {
     let author: User
     let id: UUID
     let timestamp: Date
+    var imageURL: URL? {
+        get {
+            URL(string: imageURLString)
+        }
+        set {
+            imageURLString = newValue?.absoluteString ?? ""
```
Comment (benstone1) [inline FirebaseTestProject/Models/Post.swift:21]: Should this set to an empty string if the `absoluteString` is `nil`?  Or would a `nil` value make more sense?

**Review by benstone1 — CHANGES_REQUESTED** [review]
This looks great, thanks for the quick updates!  A couple of minor notes, but this should be just about good to go.

File: FirebaseTestProject/Models/Post.swift (hunk 16 above)
Comment (john-royal) [reply FirebaseTestProject/Models/Post.swift:21]: I used an empty string because Firestore doesn’t work with `nil` values. This goes to the discussion in #8—the non-optional `imageURLString` property is a workaround to make our models compatible with Firestore.

File: FirebaseTestProject/Views/ImagePickerView.swift (hunk 12 above)
Comment (john-royal) [reply FirebaseTestProject/Views/ImagePickerView.swift:31]: Yes

File: FirebaseTestProject/Views/ProfileView.swift (hunk 15 above)
Comment (john-royal) [reply FirebaseTestProject/Views/ProfileView.swift:65]: Hard to combine this with an error alert, but I wanted this to be a `confirmationDialog` regardless. I'll go ahead and make that change.

File: FirebaseTestProject/Views/NewPostForm.swift (hunk 13 above)
Comment (john-royal) [reply FirebaseTestProject/Views/NewPostForm.swift]: Added

**Review by benstone1 — APPROVED** [review]
